### PR TITLE
Better whylabs support for logged images

### PR DESCRIPTION
**Pull request by richard-rogers**
This depends on #860 

It adds a resolver to the `ImageMetric` analogous to the `DatasetSchema`'s to create the standard metrics (counts, types, cardinality, etc.) that the visualizer & whylabs seem to require. The logical submetric name is prefixed with a metric type code so that the related metrics can be aggregated into some `ColumnProfileView` when the `ImageMetric` is "uncompounded" to send it to whylabs.

### CODE REVIEW

File: python/poetry.lock [hunk 1]
```diff
@@ -99,7 +99,7 @@ tests_no_zope = ["cloudpickle", "coverage[toml] (>=5.0.2)", "hypothesis", "mypy
 
 [[package]]
 name = "autoflake"
-version = "1.6.0"
+version = "1.6.1"
```
Comment (jamie256) [inline python/poetry.lock]: Why are you updating the poetry.lock file in this PR?

File: python/whylogs/extras/image_metric.py [hunk 2]
```diff
@@ -101,47 +105,138 @@ def image_based_metadata(img):
...
+    @abstractmethod
+    def prefixes(self) -> Set[str]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def resolve(self, name: str, why_type: DataType, fi_disabled: bool = False) -> Dict[str, Metric]:
+        raise NotImplementedError
+
+
+class StandardResolver(ImageResolver):
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py]: The name `StandardResolver` is surprisingly generic, maybe something like StandardImageMetricResolver while awkward is easier to read and understand?

File: python/whylogs/extras/image_metric.py [hunk 3]
```diff
@@ -101,47 +105,138 @@ def image_based_metadata(img):
...
+            metrics[f"fi_{name}"] = StandardMetric.frequent_items.zero(MetricConfig())
+
+        return metrics
+
+
+@dataclass(frozen=True)
+class ImageMetricConfig(MetricConfig):
+    allowed_exif_tags: Set[str] = field(default_factory=set)
+    forbidden_exif_tags: Set[str] = field(default_factory=set)
+    resolver: Resolver = field(default_factory=StandardResolver)
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py]: I feel like the class hierarchy is broken (but connected by name) here: a `StandardResolver(ImageResolver)` where the base class is `ImageResolver(ABC)` is getting assigned to a field of type `whylogs.core.resolvers.Resolver`.

Maybe having a StandardImageResolver(Resolver) would reduce the hierarchy by 1 level and simplify as you can define the `prefixes` there if needed.

File: python/whylogs/extras/image_metric.py [hunk 4]
```diff
@@ -101,47 +105,138 @@ def image_based_metadata(img):
...
-    def zero(cls, config: MetricConfig) -> "ImageMetric":
+    def zero(cls, config: Optional[MetricConfig] = None) -> "ImageMetric":
+        config = config or ImageMetricConfig()
+        if not isinstance(config, ImageMetricConfig):
+            logger.error("ImageMetric.zero() needs an ImageMetricConfig")
+            config = ImageMetricConfig()
+
         dummy_image = new_image("HSV", (1, 1))
-        metadata = get_pil_image_metadata(dummy_image)
+        metadata = get_pil_image_metadata(dummy_image)  # any EXIF tags will be discovered as images are logged
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py]: probably don't need the dummy image path if you can add to the set of things logged.

File: python/whylogs/extras/image_metric.py [hunk 5]
```diff
@@ -101,47 +105,138 @@ def image_based_metadata(img):
...
+            config.type_mapper,
+            config.fi_disabled,
+        )
 
 
 def log_image(
     images: Union[ImageType, List[ImageType], Dict[str, ImageType]],
     default_column_prefix: str = "image",
+    schema: Optional[DatasetSchema] = None,
+    default_configs: Optional[ImageMetricConfig] = None,
```
Comment (jamie256) [inline python/whylogs/extras/image_metric.py]: Adding `schema ` here as optional makes sense, but default_configs as a parameter is confusing, why doesn't the `schema.default_configs` suffice?

File: python/whylogs/api/writer/whylabs.py [hunk 6]
```diff
@@ -59,7 +59,25 @@ def _uncompounded_column_name(column_name: str, metric_name: str, submetric_name
     return f"{column_name}.{metric_name}.{submetric_name}"
 
 
+def _ugly_hack(col_name: str, metric_name: str, metric: CompoundMetric) -> Dict[str, ColumnProfileView]:
```
Comment (jamie256) [inline python/whylogs/api/writer/whylabs.py]: better would be to name this for what it is doing, and why it is needed: e.g. back_compat? image metrics backport translation? with a TODO for simplifying it if this is not needed long term.

**Review by jamie256 — CHANGES_REQUESTED** [review]
Left some questions, looks like we have something that should work for WhyLabs upload of image profiles? Can you describe what you verified in the description, or update the notebook example?

File: python/poetry.lock (hunk 1 above)
Comment (richard-rogers) [reply python/poetry.lock]: I had to update my poetry version, and some version bumps came in from mainline in the course of working on this PR.

File: python/whylogs/extras/image_metric.py (hunk 3 above)
Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: Oops, that's a bug :(  The `ImageMetricConfig::resolver` should be of type `ImageResolver`.  I looked at having `ImageResolver <: Resolver`, but the arguments to `resolve()` want to be different.

File: python/whylogs/extras/image_metric.py (hunk 4 above)
Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: That would be tricky with the allow/forbid sets. Also, SPOT for the image stats :)

File: python/whylogs/extras/image_metric.py (hunk 5 above)
Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: If don't pass a `schema`, you can specify the default config to use in the constructed schema. I would guess the more common case is that you're happy with the default `schema` but want to fiddle with the `ImageMetricConfig`.

File: python/whylogs/api/writer/whylabs.py (hunk 6 above)
Comment (richard-rogers) [reply python/whylogs/api/writer/whylabs.py]: I think writing this function convinced me we need to design something else :(

File: python/whylogs/extras/image_metric.py (hunk 5 above)
Comment (andyndang) [reply python/whylogs/extras/image_metric.py]: That's just very confusing in terms of API though. Removing API is hard so I'd err on less for now

Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: For usability, I'd prefer to drop the `schema` argument before dropping the `default_configs`.

File: python/poetry.lock (hunk 1 above)
Comment (jamie256) [reply python/poetry.lock]: but you should only have to update poetry.lock if its broken or if you change the set of dependencies in pyproject.toml file?

You didn't update the dependencies in this change, so you should not be needing to update the lock file.

Generally I think we can update the poetry.lock periodically if it hasn't been updated in a while but that is probably better as a standalone maintenance PR.

File: python/whylogs/extras/image_metric.py (hunk 4 above)
Comment (jamie256) [reply python/whylogs/extras/image_metric.py]: Can we call get_pil_image_metadata() on the first image logged and do the metric resolution then?

File: python/whylogs/extras/image_metric.py (hunk 5 above)
Comment (jamie256) [reply python/whylogs/extras/image_metric.py]: schema is consistent with the other log* APIs, I'd argue for consistency in how we extend things, even if its a little awkward to instantiate a schema just so you can throw a config in there.

File: python/poetry.lock (hunk 1 above)
Comment (richard-rogers) [reply python/poetry.lock]: It's in sync with mainline now.

File: python/whylogs/extras/image_metric.py (hunk 4 above)
Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: I restructured the code to eliminate the need for the dummy image.

File: python/whylogs/extras/image_metric.py (hunk 5 above)
Comment (richard-rogers) [reply python/whylogs/extras/image_metric.py]: Outvoted :)

**Review by jamie256 — APPROVED** [review]
LGTM!

**Pull request merged**